File: src/libs/ThirdParty.VBSuppliers/Properties/GoGlobal/GoGlobalSearch.vb
Imports System.Xml
Imports iVector.Search.Property
Imports ThirdParty.Abstractions
Imports ThirdParty.Abstractions.Models
Imports ThirdParty.Abstractions.Constants
Imports ThirdParty.Abstractions.Search.Models
Imports ThirdParty.Abstractions.Lookups
Imports Intuitive
Imports ThirdParty.VBSuppliers.My.Resources
Imports Intuitive.Net.WebRequests
Imports ThirdParty.Abstractions.Results
Imports Intuitive.Helpers.Extensions

Public Class GoGlobalSearch
    Inherits ThirdPartyPropertySearchBase

#Region "Constructor"

    Public Sub New(settings As IGoGlobalSettings, support As ITPSupport)
        _settings = Ensure.IsNotNull(settings, NameOf(settings))
        _support = Ensure.IsNotNull(support, NameOf(support))
    End Sub

#End Region

#Region "Properties"

    Private ReadOnly _settings As IGoGlobalSettings

    Private ReadOnly _support As ITPSupport

    Public Overrides Property Source As String = ThirdParties.GOGLOBAL

    Public Overrides ReadOnly Property SqlRequest As Boolean = False

    Public Overrides ReadOnly Property SupportsNonRefundableTagging As Boolean = False

#End Region

#Region "SearchRestrictions"

    Public Overrides Function SearchRestrictions(oSearchDetails As SearchDetails) As Boolean

        Dim bRestrictions As Boolean = oSearchDetails.Rooms > 1

        For Each oRoom As RoomDetail In oSearchDetails.RoomDetails
            If oRoom.ChildAges.Where(Function(i) i <= 10).Count > 2 Then bRestrictions = True
            If oRoom.Infants > 1 Then bRestrictions = True
        Next

        Return bRestrictions

    End Function

#End Region

#Region "SearchFunctions"

    Public Overrides Function BuildSearchRequests(oSearchDetails As SearchDetails, oResortSplits As List(Of ResortSplit), bSaveLogs As Boolean) As List(Of Request)

        Dim oRequests As New List(Of Request)

        'Create a request for each batch
        For Each oResortSplit As ResortSplit In oResortSplits

            'Build the request
            Dim xRequest As XDocument = BuildRequest(oSearchDetails, oResortSplit.ResortCode)

            'Make the web request
            Dim oWebRequest As New Request
            With oWebRequest
                .EndPoint = _settings.URL(oSearchDetails)
                .SetRequest(RequestWrapper("1", xRequest))
                .Method = eRequestMethod.POST
                .Source = Source
                .LogFileName = "Search"
                .TimeoutInSeconds = Me.RequestTimeOutSeconds(oSearchDetails)
                .CreateLog = bSaveLogs
                .ExtraInfo = oSearchDetails
                .UseGZip = True
            End With

            'Add the web request to the collection
            oRequests.Add(oWebRequest)
        Next

        Return oRequests

    End Function

    Public Overrides Function TransformResponse(oRequests As List(Of Request), oSearchDetails As SearchDetails, oResortSplits As List(Of ResortSplit)) As TransformedResultCollection

        Dim oTransformedXML As New XmlDocument
        Dim oXMLs As New List(Of XmlDocument)
        Dim soap As XNamespace = XNamespace.Get("http://schemas.xmlsoap.org/soap/envelope/")
        Dim ns As XNamespace = XNamespace.Get("http://www.goglobal.travel/")

        SyncLock (oRequests)
            For Each oRequest As Request In oRequests

                Dim XResponse As XDocument = XDocument.Parse(oRequest.ResponseString)

                Dim xInfo As XDocument = XDocument.Parse(XResponse.Document.Element(soap + "Envelope").Element(soap + "Body").Element(ns + "MakeRequestResponse").Element(ns + "MakeRequestResult").Value)

                Dim xmlHotel As New XmlDocument
                xmlHotel.LoadXml(xInfo.Document.Element("Root").Element("Main").ToString)

                oXMLs.Add(xmlHotel)
            Next
        End SyncLock

        'Merge the responses
        Dim oMergedResponses As XmlDocument = Intuitive.XMLFunctions.MergeXMLDocuments("Results", oXMLs)

        oTransformedXML = XMLFunctions.XMLStringTransform(oMergedResponses, GoGlobalRes.GoGlobalSearchXSL)

        Return oTransformedXML


    End Function

#End Region

#Region "Helper Functions"

    Private Function BuildRequest(ByVal oSearchDetails As SearchDetails, ByVal CityID As String) As XDocument

        Dim xRequest As XDocument = New XDocument(New XElement("Root",
                                                               New XElement("Header",
                                                                            New XElement("Agency", _settings.Agency(oSearchDetails)),
                                                                            New XElement("User", _settings.User(oSearchDetails)),
                                                                            New XElement("Password", _settings.Password(oSearchDetails)),
                                                                            New XElement("Operation", "HOTEL_SEARCH_REQUEST"),
                                                                            New XElement("OperationType", "Request")
                                                                            ),
                                                                New XElement("Main",
                                                                             New XElement("MaximumWaitTime", Me.RequestTimeOutSeconds(oSearchDetails)),
                                                                             New XElement("CityCode", CityID),
                                                                             New XElement("ArrivalDate", oSearchDetails.ArrivalDate.ToString("yyyy-MM-dd")),
                                                                             New XElement("Nights", oSearchDetails.Duration),
                                                                             New XElement("Rooms",
                                                                                          From oRoom In oSearchDetails.RoomDetails
                                                                                          Select New XElement("Room",
                                                                                                New XAttribute("Adults", oRoom.Adults + oRoom.ChildAges.Where(Function(i) i > 10).Count),
                                                                                                New XAttribute("RoomCount", oSearchDetails.Rooms),
                                                                                                New XAttribute("CotCount", oRoom.Infants),
                                                                                                From iChildAge In oRoom.ChildAges
                                                                                                Where iChildAge <= 10
                                                                                                Select New XElement("ChildAge", iChildAge))))))
        Return xRequest


    End Function

    Private Function RequestWrapper(ByVal RequestType As String, ByVal XML As XDocument) As String

        Dim xsi As XNamespace = XNamespace.Get("http://www.w3.org/2001/XMLSchema-instance")
        Dim xsd As XNamespace = XNamespace.Get("http://www.w3.org/2001/XMLSchema")
        Dim soap As XNamespace = XNamespace.Get("http://schemas.xmlsoap.org/soap/envelope/")
        Dim ns As XNamespace = XNamespace.Get("http://www.goglobal.travel/")

        Dim xRequest As XDocument = New XDocument(
                                     New XDeclaration("1.0", "utf-8", "no"),
                                     New XElement(soap + "Envelope",
                                                  New XAttribute(XNamespace.Xmlns + "xsi", xsi),
                                                  New XAttribute(XNamespace.Xmlns + "xsd", xsd),
                                                  New XAttribute(XNamespace.Xmlns + "soap", soap),
                                                  New XElement(soap + "Body",
                                                               New XElement(ns + "MakeRequest",
                                                                            New XAttribute("xmlns", ns),
                                                                            New XElement(ns + "requestType", RequestType),
                                                                            New XElement(ns + "xmlRequest", New XCData(XML.ToString))))))

        Return xRequest.ToString

    End Function

    Private Function RoomCodes(ByVal Room As RoomDetail) As Generic.List(Of String)

        Dim iOccupancy As Integer = Room.Adults + Room.ChildAges.Where(Function(i) i > 10).Count

        Dim sCodes As New Generic.List(Of String)

        If iOccupancy = 1 Then sCodes = New Generic.List(Of String) From {"SGL", "DBLSGL"}
        If iOccupancy = 2 Then sCodes = New Generic.List(Of String) From {"DBL", "TWN"}
        If iOccupancy = 3 Then sCodes.Add("TPL")
        If iOccupancy = 4 Then sCodes.Add("QDR")

        Return sCodes

    End Function

#End Region

#Region "ResponseHasExceptions"

    Public Overrides Function ResponseHasExceptions(oRequest As Intuitive.Net.WebRequests.Request) As Boolean
        Return False
    End Function

#End Region

    Public Class GoGlobalSearchHelper
        Public Property SearchDetails As SearchDetails
        Public Property RoomTypes As New Generic.List(Of String)

        Public Sub New(ByVal Details As SearchDetails, ByVal Types As Generic.List(Of String))
            SearchDetails = Details
            RoomTypes = Types
        End Sub

        Public Sub New(ByVal Details As SearchDetails)
            SearchDetails = Details
        End Sub

    End Class

    Private Class Room
        Public Property Adults As Integer
        Public Property RoomCount As Integer
        Public Property ChildAges As New Generic.List(Of Integer)
        Public Property CotCount As Integer

        Public Sub New()

        End Sub

        Public Sub New(Adults As Integer, ChildAges As String, CotCount As Integer, RoomCount As Integer)
            Me.Adults = Adults
            Me.RoomCount = RoomCount
            Me.CotCount = CotCount

            For Each sAge As String In ChildAges.Split(","c)
                If sAge.ToSafeInt() <= 10 Then
                    Me.ChildAges.Add(sAge.ToSafeInt())
                Else
                    Me.Adults += 1
                End If
            Next
        End Sub
    End Class
End Class

File: src/libs/iVectorOne.Suppliers/GoGlobal/GoGlobalSearch.cs
namespace iVectorOne.Suppliers.GoGlobal
{
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;
    using Intuitive;
    using Intuitive.Helpers.Extensions;
    using Intuitive.Helpers.Net;
    using Intuitive.Helpers.Serialization;
    using iVectorOne.Constants;
    using iVectorOne.Interfaces;
    using iVectorOne.Models;
    using iVectorOne.Models.Property.Booking;
    using iVectorOne.Search.Models;
    using iVectorOne.Search.Results.Models;
    using iVectorOne.Suppliers.GoGlobal.Models;
    using Cancellation = iVectorOne.Models.Cancellation;
    using Newtonsoft.Json;
    using iVectorOne.Models.Property;

    public partial class GoGlobalSearch : IThirdPartySearch, ISingleSource
    {
        #region "Properties"

        private IGoGlobalSettings _settings;
        private readonly ISerializer _serializer;

        public string Source => ThirdParties.GOGLOBAL;

        #endregion

        #region "Constructors"

        public GoGlobalSearch(IGoGlobalSettings settings, ISerializer serializer)
        {
            _settings = Ensure.IsNotNull(settings, nameof(settings));
            _serializer = Ensure.IsNotNull(serializer, nameof(serializer));
        }

        #endregion

        #region "SearchRestrictions"

        public bool SearchRestrictions(SearchDetails oSearchDetails, string source)
        {
            if (oSearchDetails.Rooms > 1) return true;

            // Supplier Restrictions
            // ALL rooms in a request must have Adults
            // The maximum amount of persons (adults + children), per room, is 8
            // The limit of children, per room, is 4
            // Children are between ages 1 to 18 when using version 2.2 and above(Legacy support ages 2 to 10
            bool bRestricted = oSearchDetails.RoomDetails.Any(oRoom =>
            {
                if (oRoom.Adults == 0) return true;
                if (oRoom.Adults + oRoom.Children > 8) return true;
                if (oRoom.ChildAges.Any(age => age < 1 && age > 18)) return true;
                return false;
            });

            return bRestricted;
        }

        #endregion

        #region "SearchFunctions"

        public Task<List<Request>> BuildSearchRequestsAsync(SearchDetails oSearchDetails, List<ResortSplit> oResortSplits)
        {
            var oRequests = new List<Request>();

            string sNationalityCode = string.IsNullOrEmpty(oSearchDetails.ISONationalityCode)
                ? _settings.LeadGuestNationality(oSearchDetails)
                : oSearchDetails.ISONationalityCode;

            var TPKeyBatches = oResortSplits.SelectMany(split => split.Hotels.Select(h => h.TPKey))
                                .Distinct().Batch(_settings.HotelBatchLimit(oSearchDetails)).ToList();

            foreach (var TPKeyBatch in TPKeyBatches)
            {

                int idxRoom = 0;
                foreach (var oRoom in oSearchDetails.RoomDetails)
                {
                    idxRoom++;

                    var infantChild = Enumerable.Range(0, oRoom.Infants).Select(_ => "1");
                    var childAges = oRoom.ChildAges.Select(age => $"{age}").Concat(infantChild).ToList();

                    //'Build the request
                    var oRequestData = new SearchRq
                    {

                        Version = Constant.Version_2_3,
                        ResponseFormat = Constant.FormatJSON,
                        Hotels = TPKeyBatch.ToList(),
                        Nationality = sNationalityCode,
                        ArrivalDate = oSearchDetails.ArrivalDate.ToString(Constant.DataFormat),
                        Nights = oSearchDetails.Duration.ToString(),
                        Rooms =
                        {
                            new Room
                            {
                                Adults = $"{oRoom.Adults + oRoom.ChildAges.Count(age => age > Constant.MaxChildAge)}",
                                RoomCount = $"1",
                                ChildCount = $"{oRoom.Infants + oRoom.Children}",
                                ChildAges = oRoom.ChildAges.Where(age => age <= Constant.MaxChildAge).Select(age => $"{age}")
                                            .Concat(Enumerable.Range(0, oRoom.Infants).Select(i => "1")).ToList()
                            }
                        },
                    };

                    var oRoot = new Root<SearchRq>
                    {
                        Header =
                        {
                             Agency = _settings.Agency(oSearchDetails),
                             User = _settings.User(oSearchDetails),
                             Password = _settings.Password(oSearchDetails),
                             Operation = Constant.RequestHotelSearch,
                             OperationType = Constant.OperationTypeRequest,
                        },
                        Main = oRequestData
                    };

                    string sRoot = $"{_serializer.Serialize(oRoot).OuterXml}";
                    var envelope = Envelope.CreateRequest(Constant.RequestCodeHotelSearch, sRoot, _serializer);

                    //'Make the web request
                    var oWebRequest = new Request
                    {
                        EndPoint = _settings.GenericURL(oSearchDetails),
                        Method = RequestMethod.POST,
                        Source = Source,
                        LogFileName = "Search",
                        CreateLog = true,
                        ExtraInfo = $"{idxRoom}",
                        UseGZip = true,
                    };
                    oWebRequest.SetRequest(envelope);

                    oRequests.Add(oWebRequest);
                }
            }

            return Task.FromResult(oRequests);
        }

        public TransformedResultCollection TransformResponse(List<Request> oRequests, SearchDetails oSearchDetails, List<ResortSplit> oResortSplits)
        {
            var transformedResults = new TransformedResultCollection();

            foreach (var oRequest in oRequests)
            {
                string response = Envelope.GetResponse(oRequest.ResponseXML, _serializer);
                var searchResult = JsonConvert.DeserializeObject<SearchRs>(response);
                int roomIdx = oRequest.ExtraInfo.ToSafeInt();

                var transfomedRooms = searchResult.Hotels.SelectMany(oHotel =>
                {
                    return oHotel.Offers.SelectMany(offer =>
                    {
                        decimal nRate = offer.TotalPrice.ToSafeDecimal();
                        var cancellationList = new Cancellations();

                        var cancellations = GoGlobal.GetCancellationPeriods(offer.Remark, oSearchDetails.ArrivalDate)
                                .Select(c => new Cancellation
                                {
                                    StartDate = c.StartDate,
                                    EndDate = c.EndDate.AddDays(-1),
                                    Amount = nRate * c.FeePercent / 100
                                });

                        cancellationList.AddRange(cancellations);

                        return offer.Rooms.Select((sRoomType, idx) =>
                        {
                            return new TransformedResult
                            {
                                MasterID = oHotel.HotelCode.ToSafeInt(),
                                PropertyRoomBookingID = roomIdx,
                                MealBasisCode = offer.RoomBasis,
                                TPKey = oHotel.HotelCode,
                                CurrencyCode = offer.Currency,
                                RoomType = sRoomType,
                                Amount = nRate,
                                TPReference = $"{offer.HotelSearchCode}",
                                NonRefundableRates = offer.NonRef,
                                Cancellations = cancellationList
                            };
                        });
                    });
                }).ToList();
                transformedResults.TransformedResults.AddRange(transfomedRooms);
            }
            
            return transformedResults;
        }

        #endregion

        #region "ResponseHasExceptions"

        public bool ResponseHasExceptions(Request request)
        {
            return false;
        }

        #endregion
    }
}
